BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m

```diff
@@ -0,0 +1,297 @@
+{
+    BOXURLBackgroundSessionIdAndSessionTaskId *backgroundSessionIdAndSessionTaskId = [self backgroundSessionIdAndSessionTaskIdGivenUserId:userId associateId:associateId];
+    if (backgroundSessionIdAndSessionTaskId == nil) {
+        return YES;
+    }
+
+    NSString *backgroundSessionId = backgroundSessionIdAndSessionTaskId.backgroundSessionId;
+    NSUInteger sessionTaskId = backgroundSessionIdAndSessionTaskId.sessionTaskId;
+
+    //clean up sessions/$backgroundSessionId/$sessionTaskId/*
+    NSString *dir = [self dirPathOfSessionTaskWithBackgroundSessionId:backgroundSessionId sessionTaskId:sessionTaskId];
+    BOOL success = [self deleteDirectory:dir error:error];
+
+    if (success == YES) {
+        //clean up users/$userId/$associateId/*
+        dir = [self dirPathOfUserSessionTaskGivenUserId:userId associateId:associateId];
+        success = [self deleteDirectory:dir error:error];
+    }
+    return success;
+}
+
+#pragma mark - private helpers
+
+- (BOOL)cacheBackgroundSessionId:(NSString *)backgroundSessionId sessionTaskId:(NSUInteger)sessionTaskId data:(NSData *)data type:(BOXURLSessionTaskCacheFileType)type error:(NSError **)outError
+{
+    NSError *error;
+
+    //persist sessions/$backgroundSessionId/$sessionTaskId
+    BOOL success = [self createDirForBackgroundSessionId:backgroundSessionId sessionTaskId:sessionTaskId error:&error];
+
+    if (success == YES && error == nil) {
+        NSString *path = [self dirPathOfSessionTaskWithBackgroundSessionId:backgroundSessionId sessionTaskId:sessionTaskId];
+        //persist data to sessions/$backgroundSessionId/$sessionTaskId/$fileType
+
+        switch (type) {
+            case BOXURLSessionTaskCacheFileTypeDestinationFilePath:
+                path = [path stringByAppendingPathComponent:BOXURLSessionTaskCacheDestinationFilePath];
+                break;
+            case BOXURLSessionTaskCacheFileTypeResumeData:
+                path = [path stringByAppendingPathComponent:BOXURLSessionTaskCacheResumeData];
+                break;
+            case BOXURLSessionTaskCacheFileTypeResponse:
+                path = [path stringByAppendingPathComponent:BOXURLSessionTaskCacheResponse];
+                break;
+            case BOXURLSessionTaskCacheFileTypeResponseData:
+                path = [path stringByAppendingPathComponent:BOXURLSessionTaskCacheResponseData];
+                break;
+            case BOXURLSessionTaskCacheFileTypeError:
+                path = [path stringByAppendingPathComponent:BOXURLSessionTaskCacheError];
+                break;
+        }
+
+        success = [self cacheAndEncryptData:data atPath:path error:&error];
+    }
+
+    if (outError != nil) {
+        (*outError) = error;
+    }
+    return success;
+}
+
+- (BOOL)cacheAndEncryptData:(NSData *)data atPath:(NSString *)path error:(NSError **)outError
+{
+    NSData *finalData = data;
+    if ([self.delegate respondsToSelector:@selector(encryptData:)]) {
```

Comment:
If the delegate doesn't respond to the selector, can we really say that we encrypted the file? Might want to reflect an "attempt" abstraction in the method signature if this isn't always the case.